lib/ansible/plugins/connection/network_cli.py

```diff
@@ -380,9 +380,6 @@ def receive(self, command=None, prompts=None, answer=None, newline=True, prompt_
         command_prompt_matched = False
         matched_prompt_window = window_count = 0
 
-        command_timeout = self.get_option('persistent_command_timeout')
```

Comment:
Removing the signal from here won't entirely solve the issue as `persistent_command_timeout` timer signal is also triggered from `ansible-connection` which is common for all the connection plugins using the persistent framework (netconf/network_cli/httpapi). This mix of signal and sleep in `network_cli` is resulting in non-deterministic behavior, based on this I think we need to revert PR https://github.com/ansible/ansible/pull/47345 as the signal based timeout handling is in place since 2.3 and is working fine.

Reply:
> it does not happen under normal conditions or during regular network activity and it has a very long delay
`persistent_command_timeout` default timeout value triggered from `ansible-connection` is 10 secs, not sure how you concluded it is a 'very long delay'.

Reply:
By very long delay, I was meaning relative to the other signal delay of 100ms. But once again, since persistent_command_timeout triggers the end of program, I dont see the issue here.

Also, I'm not sure to understand the sleep issue. Before #47345, you were signaling around paramiko's recv, which will be sleeping nearly all the time when signal handler is called. I have read signal may not interrupts a sleep call and may triggers only when it has finished. But who cares when sleep just lasting few milliseconds ? Is this about this behavior or am i missing something here ?

In summary, in my opinion, the issue is more related to finally clauses and using signal with exceptions. When paramiko is already in the middle of a finally clause, cleaning everything up (like going to release a lock), when at that exact time the signal handler is called, since it raises an exception the finally clause can't finish her work and leaves in a dirty state (like with a lock acquired). Don't you agree that's the issue's root ?

Reply:
The 0.001 sleep doesn't seem to to be sufficient for recv_ready() and after couple of runs results in an deadlock. 
Looking at performance improvement with this check it seems it not that significant at scale unless I am missing something. Considering 2sec/per command (optimistic estimation) for 1000 command will take around ~33.33 mins to complete playbook run. With this check added considering the best improvement estimate will be 1.95 sec/per command that is ~32.5 mins.

The sleep in recv() doesn't result in deadlock, even if the `persistent_command_timeout` signal handler is missed it is bailed out from while loop by `socket.timeout` exeption handling.

Reply:
> The 0.001 sleep doesn't seem to to be sufficient for recv_ready() and after couple of runs results in an deadlock.
This has little to do with sleep length. Increasing sleep time just lowering probability that a finally clause get destroyed by the exception raised in signal handler, but just because you will run less finally clauses.
 
> Looking at performance improvement with this check it seems it not that significant at scale unless I am missing something. Considering 2sec/per command (optimistic estimation) for 1000 command will take around ~33.33 mins to complete playbook run. With this check added considering the best improvement estimate will be 1.95 sec/per command that is ~32.5 mins.

2 seconds per command ? On my cheap network hardware, a command typically ran at 100ms during my tests (including ~55ms of ssh time without #47345). When you declare a vlan on it, it sets all the switch interfaces on dynamically registered through GVRP, so you have to exclude the vlan everywhere you don't want it to have well defined vlans. With 48 ports switchs and some vlans, it generates a lot of commands. Gaining 25% is not negligeable on that part.
 
> The sleep in recv() doesn't result in deadlock, even if the `persistent_command_timeout` signal handler is missed it is bailed out from while loop by `socket.timeout` execption handling.

Recv also has a finally clause that is also releasing a lock just like recv_ready. And therefore it can cause a deadlock like recv_ready with signals raising exceptions. It's just a lot less likely to happens because there is only one recv call (and one finally clause run) for every signal trigger.